### PR TITLE
add ImageData class

### PR DESCRIPTION
**Pull request by vincentsarago**
continuation of #297 but only for the `Image` part

### CODE REVIEW

File: rio_tiler/models.py [hunk 1]
```diff
@@ -67,3 +71,77 @@ class Metadata(Info):
...
+
+@attr.s
+class ImageData:
+    """Image Data class."""
+
+    data: numpy.ndarray = attr.ib()
+    mask: numpy.ndarray = attr.ib()
+    assets: Optional[List[str]] = attr.ib(default=None)
+
+    def __iter__(self):
```
Comment (geospatial-jeff) [inline rio_tiler/models.py:84]: nice

**Review by geospatial-jeff — APPROVED** [review]

File: rio_tiler/mosaic/reader.py [hunk 2]
```diff
@@ -83,15 +82,10 @@ def reader(asset: str, *args, **kwargs) -> Tuple[numpy.ndarray, numpy.ndarray]:
...
+        for img, asset in filter_tasks(tasks, allowed_exceptions=allowed_exceptions,):
             assets_used.append(asset)
-            t = numpy.ma.array(t)
-            t.mask = m == 0
-
-            pixel_selection.feed(t)
+            pixel_selection.feed(img.as_masked())
             if pixel_selection.is_done:
-                return pixel_selection.data, assets_used
+                return ImageData(*pixel_selection.data, assets=assets_used), assets_used  # type: ignore
```
Comment (vincentsarago) [inline rio_tiler/mosaic/reader.py:89]: note sure why mypy is complaining here about ImageData receiving assets twice 🤷‍♂️

Comment (vincentsarago) [inline rio_tiler/mosaic/reader.py:89]: We return `ImageData, assets_used` be backward compatible.

Comment (vincentsarago) [inline rio_tiler/mosaic/reader.py:89]: `ImageData`. doesn't have the `crs` nor `bounds` for the `reader` result. I guess we could forward it from the last `img`

File: rio_tiler/utils.py [hunk 3]
```diff
@@ -253,27 +253,33 @@ def non_alpha_indexes(src_dst: Union[DatasetReader, DatasetWriter, WarpedVRT]) -
 
 def linear_rescale(
     image: numpy.ndarray,
-    in_range: Tuple[Union[int, float], Union[int, float]] = (0, 1),
-    out_range: Tuple[Union[int, float], Union[int, float]] = (1, 255),
+    in_range: Tuple[NumType, NumType],
+    out_range: Optional[Tuple[NumType, NumType]] = None,
```
Comment (vincentsarago) [inline rio_tiler/utils.py:257]: cc @kylebarron (ref: https://github.com/cogeotiff/rio-tiler/pull/297#discussion_r515162342)

File: rio_tiler/mosaic/reader.py [hunk 4]
```diff
@@ -83,15 +82,13 @@ def reader(asset: str, *args, **kwargs) -> Tuple[numpy.ndarray, numpy.ndarray]:
...
         tasks = create_tasks(reader, chunks, threads, *args, **kwargs)
-        for (t, m), asset in filter_tasks(
-            tasks, allowed_exceptions=allowed_exceptions,
-        ):
-            assets_used.append(asset)
-            t = numpy.ma.array(t)
-            t.mask = m == 0
+        for img, asset in filter_tasks(tasks, allowed_exceptions=allowed_exceptions,):
+            if isinstance(img, tuple):
+                img = ImageData(*img)
```
Comment (vincentsarago) [inline rio_tiler/mosaic/reader.py:87]: Keeping compatibility with old style `reader` definition
```python
def _read_preview(
    src_path: str, *args, **kwargs
) -> Tuple[numpy.ndarray, numpy.ndarray]:
    """Read preview from an asset"""
    with COGReader(src_path) as cog:
        data, mask = cog.preview(*args, **kwargs)
    return data, mask
```

File: rio_tiler/constants.py [hunk 5]
```diff
@@ -2,10 +2,15 @@
 
 import multiprocessing
 import os
+from typing import Tuple, Union
 
 import morecantile
 from rasterio.crs import CRS
 
+NumType = Union[float, int]
+BBox = Tuple[float, float, float, float]
```
Comment (vincentsarago) [inline rio_tiler/constants.py:11]: cc @kylebarron

File: rio_tiler/mosaic/reader.py [hunk 6]
```diff
@@ -81,17 +82,22 @@ def reader(asset: str, *args, **kwargs) -> Tuple[numpy.ndarray, numpy.ndarray]:
...
         tasks = create_tasks(reader, chunks, threads, *args, **kwargs)
-        for (t, m), asset in filter_tasks(
-            tasks, allowed_exceptions=allowed_exceptions,
-        ):
+        for img, asset in filter_tasks(tasks, allowed_exceptions=allowed_exceptions,):
+            if isinstance(img, tuple):
+                img = ImageData(*img)
+
+            crs = img.crs
+            bounds = img.bounds
```
Comment (vincentsarago) [inline rio_tiler/mosaic/reader.py:95]: we are always returning the last crs/bounds returned!

File: rio_tiler/models.py [hunk 7]
```diff
@@ -75,3 +82,108 @@ class Metadata(Info):
...
+
+@attr.s
+class ImageData:
+    """Image Data class."""
+
+    data: numpy.ndarray = attr.ib()
+    mask: numpy.ndarray = attr.ib()
+    assets: Optional[List[str]] = attr.ib(default=None)
+    bounds: Optional[BBox] = attr.ib(default=None)
+    crs: Optional[CRS] = attr.ib(default=None)
```
Comment (vincentsarago) [inline rio_tiler/models.py:95]: any other properties? cc @kylebarron @geospatial-jeff

File: rio_tiler/models.py [hunk 8]
```diff
@@ -75,3 +82,108 @@ class Metadata(Info):
...
+            if "transform" not in kwargs:
+                kwargs.update({"transform": self.transform})
+            if "crs" not in kwargs and self.crs:
+                kwargs.update({"crs": self.crs})
+
+        return (
+            render(self.data, self.mask, img_format=img_format, **kwargs)
+            if add_mask
+            else render(self.data, img_format=img_format, **kwargs)
+        )
```
Comment (kylebarron) [inline rio_tiler/models.py:189]: My preference for readability in these situations is 

```suggestion
        if add_mask:
            return render(self.data, self.mask, img_format=img_format, **kwargs)

        return render(self.data, img_format=img_format, **kwargs)
```

but not a huge deal

File: rio_tiler/constants.py (hunk 5 above)
Comment (kylebarron) [reply rio_tiler/constants.py:11]: I see you use it in post process

File: rio_tiler/utils.py [hunk 9]
```diff
@@ -253,27 +253,33 @@ def non_alpha_indexes(src_dst: Union[DatasetReader, DatasetWriter, WarpedVRT]) -
...
-            returns rescaled image array.
+    out: numpy ndarray
+        returns rescaled image array.
 
     """
+    if out_range is None:
+        warnings.warn(
+            "No output range set, using image (0, 255)", UserWarning,
+        )
+        out_range = (0, 255)
```
Comment (kylebarron) [inline rio_tiler/utils.py:281]: Up to you whether you want to raise this warning. I don't have a preference if the default behavior is just to rescale to 0-255. Makes sense for most use cases.

File: setup.py [hunk 10]
```diff
@@ -15,6 +15,7 @@
     "morecantile>=1.1,<2.0",
     "rasterio>=1.1.7",
     "requests",
+    "rio-color",
```
Comment (kylebarron) [inline setup.py:18]: Ref https://github.com/mapbox/rio-color/pull/66

**Review by kylebarron — APPROVED** [review]

File: rio_tiler/io/base.py [hunk 11]
```diff
@@ -315,9 +306,9 @@ def _reader(
 
         if expression:
             blocks = expression.split(",")
-            data = apply_expression(blocks, assets, data)
+            output.data = apply_expression(blocks, assets, output.data)
```
[the remark this answers is not in the document]
Comment (vincentsarago) [reply rio_tiler/io/base.py:309]: yes it is!

File: rio_tiler/io/cogeo.py [hunk 12]
```diff
@@ -401,14 +403,24 @@ def part(
...
 
-        return data, mask
+        return ImageData(
+            data,
+            mask,
+            bounds=(
+                bbox
+                if dst_crs == bounds_crs
+                else transform_bounds(bounds_crs, dst_crs, *bbox, densify_pts=21)
+            ),
```
[the remark this answers is not in the document]
Comment (vincentsarago) [reply rio_tiler/io/cogeo.py:413]: we use this function in https://github.com/cogeotiff/rio-tiler/pull/298/files#diff-ef234b256a29630c9fec85144eaf6a6689f001129d364af8fe82a23b73675bf9R181 so to me it made more sense to directly add it inside the object especially if bounds and CRS are defined in the object.

File: rio_tiler/models.py [hunk 13]
```diff
@@ -75,3 +82,108 @@ class Metadata(Info):
...
+    bounds: Optional[BBox] = attr.ib(default=None)
+    crs: Optional[CRS] = attr.ib(default=None)
+
+    def __iter__(self):
+        """Allow for variable expansion (``arr, mask = ImageData``)"""
+        for i in (self.data, self.mask):
+            yield i
+
+    @classmethod
+    def create_from_list(cls, data: Sequence):
```
[the remark this answers is not in the document]
Comment (vincentsarago) [reply rio_tiler/models.py:103]: yeah I tried `Sequence[ImageData]` and it doesn't work, I'll try with `'ImageData'`

[the remark this answers is not in the document]
Comment (kylebarron) [reply rio_tiler/models.py:103]: https://github.com/python/mypy/issues/3661#issuecomment-313157498

File: rio_tiler/utils.py (hunk 9 above)
Comment (vincentsarago) [reply rio_tiler/utils.py:281]: because we don't add `out_range=(0, 255)` in the function definition I think it's important to add this warning.

Comment (kylebarron) [reply rio_tiler/utils.py:281]: Yeah I'm saying that I don't have a preference between how it is now and removing the warning and including `out_range=(0, 255)` as the default value in the def

File: rio_tiler/io/cogeo.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kylebarron) [reply rio_tiler/io/cogeo.py:413]: I meant for readability something like:

```py
if dst_crs == bounds_crs:
    bounds = bbox
else:
    bounds = transform_bounds(bounds_crs, dst_crs, *bbox, densify_pts=21)

return ImageData(
    data,
    mask,
    bounds=bounds,
    crs=dst_crs,
    assets=[self.filepath],
)
```

[the remark this answers is not in the document]
Comment (vincentsarago) [reply rio_tiler/io/cogeo.py:413]: 🤦

File: rio_tiler/models.py [hunk 14]
```diff
@@ -75,3 +82,111 @@ class Metadata(Info):
...
+    @property
+    def transform(self):
+        """Returns the affine transform."""
+        return (
+            from_bounds(*self.bounds, self.width, self.height)
+            if self.bounds
+            else Affine.scale(self.width, -self.height)
+        )
+
+    def post_process(
```
Comment (vincentsarago) [inline rio_tiler/models.py:155]: @kylebarron @geospatial-jeff maybe this method should return a new instance?

File: rio_tiler/models.py [hunk 15]
```diff
@@ -75,3 +82,111 @@ class Metadata(Info):
...
+                    linear_rescale(
+                        self.data[bdx], in_range=rescale_arr[bdx], out_range=out_range
+                    ),
+                    0,
+                )
+
+        if color_formula:
+            self.data[self.data < 0] = 0
+            for ops in parse_operations(color_formula):
+                self.data = scale_dtype(ops(to_math_type(self.data)), numpy.uint8)
```
Comment (vincentsarago) [inline rio_tiler/models.py:179]:
```suggestion
        inplace: bool = False,
    ) -> 'ImageData':
        """Post-process image data."""
        data = self.data if inplace else self.data.copy()
        mask = self.mask if inplace else self.mask.copy()

        if in_range:
            rescale_arr = tuple(_chunks(in_range, 2))
            if len(rescale_arr) != self.count:
                rescale_arr = ((rescale_arr[0]),) * self.count

            for bdx in range(self.count):
                data[bdx] = numpy.where(
                    self.mask,
                    linear_rescale(
                        data[bdx], in_range=rescale_arr[bdx], out_range=out_range
                    ),
                    0,
                )

        if color_formula:
            data[data < 0] = 0
            for ops in parse_operations(color_formula):
                data = scale_dtype(ops(to_math_type(data)), numpy.uint8)

        return self if inplace else ImageData(data, mask, crs=self.crs, bounds=self.bounds, assets=self.assets)
```

Comment (vincentsarago) [reply rio_tiler/models.py:179]: I think it makes sense to add the `inplace` method to let the user decide if they want to overwrite the object or create a new one!

**Pull request merged**